fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -48,46 +50,121 @@ impl<T> Interpreter<PredicateStorage, T> {
         if !checked.transaction().check_predicate_owners(&params) {
             return Err(PredicateVerificationFailed::InvalidOwner);
         }
 
         let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
 
-        // Needed for now because checked is only freed once the value is collected into a Vec
-        #[allow(clippy::needless_collect)]
-        let predicates: Vec<_> = (0..checked.transaction().inputs().len())
-            .filter_map(|i| RuntimePredicate::from_tx(&params, checked.transaction(), i))
-            .collect();
+        let mut cumulative_gas_used: Word = 0;
 
-        // Since we reuse the vm objects otherwise, we need to keep the actual gas here
-        let tx_gas_limit = checked.transaction().limit();
-        let mut remaining_gas = tx_gas_limit;
+        vm.init_predicate(checked.clone());
 
-        vm.init_predicate(checked);
+        for (idx, input) in checked.transaction().inputs().iter().enumerate().filter(|(_, input)| {
+            matches!(
+                input,
+                Input::CoinPredicate(_) | Input::MessageCoinPredicate(_) | Input::MessageDataPredicate(_)
+            )
+        }) {
+            if let Some(predicate) = RuntimePredicate::from_tx(&params, checked.transaction(), idx) {
+                // VM is cloned because the state should be reset for every predicate verification
+                let mut vm = vm.clone();
 
-        for predicate in predicates {
-            // VM is cloned because the state should be reset for every predicate verification
-            let mut vm = vm.clone();
+                vm.context = Context::PredicateVerification {
+                    program: predicate.clone(),
+                };
 
-            vm.context = Context::Predicate { program: predicate };
-            vm.set_remaining_gas(remaining_gas);
+                let gas_used = if let Some(x) = input.predicate_gas_used() {
+                    x
+                } else {
+                    return Err(PredicateVerificationFailed::GasNotSpecified);
+                };
+
+                vm.set_gas(gas_used);
 
-            if !matches!(vm.verify_predicate()?, ProgramState::Return(0x01)) {
-                return Err(PredicateVerificationFailed::False);
+                if !matches!(vm.verify_predicate()?, ProgramState::Return(0x01)) {
+                    return Err(PredicateVerificationFailed::False);
+                }
+
+                if vm.registers[RegId::GGAS] != 0 {
+                    return Err(PredicateVerificationFailed::GasMismatch);
+                }
+                cumulative_gas_used = cumulative_gas_used
+                    .checked_add(gas_used)
+                    .ok_or_else(|| PredicateVerificationFailed::OutOfGas)?;
             }
+        }
 
-            remaining_gas = vm.registers[RegId::GGAS];
+        if cumulative_gas_used > checked.transaction().limit() {
+            return Err(PredicateVerificationFailed::OutOfGas);
```

Comment:
I think we want to return another error here, more user-friendly

Reply:
added CumulativePredicateGasExceededTxGasLimit